### PR TITLE
Support binary symbol names and unicode

### PR DESCRIPTION
**Pull request by Leont**
This combines the tests from #5 with fixes of my own to make Sub::Name both Unicode compatible and binary safe on perls that support that (within rules that perl has for names: `::` and `'` are treated as namespace separators, so they can't be otherwise used in the name).

This PR depends on mhx/Devel-PPPort#32 being released first, without that it will fail for lack of gv_init_pvn (5.10 and lower) and gv_fetchpvn_flags (5.8 and lower)

### CODE REVIEW

**Comment by karenetheridge** [comment]
note to self: we'll need a develop-requires dep on the new Devel::PPPort to make sure that it gets into the Sub-Name release.

File: t/exotic_names.t [hunk 1]
```diff
@@ -59,7 +59,7 @@ sub caller3_ok {
 
 use Sub::Name 'subname';
 
-my @ordinal = ( 1 .. 255 );
+my @ordinal = ( 1 .. 38, 40 .. 255 );
```
Comment (ribasushi) [inline t/exotic_names.t]: Commenting so things do not fall through the cracks: @Leont this commit needs to be reverted, and the XS-guts amended to pass as per the conversation on March 16th:
```
[01:44:11] <leont> ribasushi, aristotle: your opinions on apostrophe's in Sub::Name wasn't entirely clear to me
[01:45:24] <leont> Sub::Name always treated them as separators, but the tests suggest it should be treated as a normal character.
[01:46:13] <leont> Given how much perl's stash interfaces all resist this, I don't think this is sensible
[01:56:21] <aristotle> you can invoke a function whose name is not a valid identifier by symbolic reference
[01:56:44] <aristotle> but you cannot do that for a function whose name contains a literal :: or '
[01:57:09] <aristotle> because Perl always takes those to be package separators when it looks up the symref
[01:57:39] <aristotle> consequently I think Sub::Name should consider them package separators also
[01:57:54] <aristotle> does that make sense?
[01:57:58] <leont> Yes
```

**Comment by ribasushi** [comment]
Not sure why GH considered [my comment](https://github.com/p5sagit/Sub-Name/pull/8#discussion_r67602791) to have been made "on outdated diff": what I was referring to is [this line](https://github.com/p5sagit/Sub-Name/pull/8/files#diff-a99b09603d90c33ce955fe8a7ef3343bR62) needing to **include** `39/x27`.

**Comment by Leont** [comment]
> Not sure why GH considered my comment to have been made "on outdated diff": what I was referring to is this line needing to include 39/x27.

I'm not sure what's going on there. At first I thought the test was wrong (hence me removing it), but on further inspection it seems the test is sensible but the results it's getting isn't:

~~~~
Use of uninitialized value in join or string at t/exotic_names.t line 52.
not ok 157 - stash name for renamed closure is correct when it contains \x27 ( ' )
#   Failed test 'stash name for renamed closure is correct when it contains \x27 ( ' )'
#   at t/exotic_names.t line 89.
#          got: '::_NAME'
#     expected: 'test::SOME_::_STASH::SOME_::_NAME'
not ok 158 - caller() in renamed closure returns correct name when it contains \x27 ( ' )
#   Failed test 'caller() in renamed closure returns correct name when it contains \x27 ( ' )'
#   at t/exotic_names.t line 89.
#          got: '__ANON__::_NAME'
#     expected: 'test::SOME_::_STASH::SOME_::_NAME'
~~~~

**Comment by ribasushi** [comment]
> on further inspection it seems the test is sensible but the results it's getting isn't

Thus me saying earlier

> ...and the XS-guts amended to pass

In other words: the PR is **not ready** ;)

**Comment by Leont** [comment]
Now it is ;-)

**Comment by Leont** [comment]
Now it is ;-)

**Comment by ribasushi** [comment]
@ap It would be really beneficial if you comment on my thinking in the P.S. below

@leont The code as seen in e82bf3b1 is sufficiently correct for my needs, awesome work!
It does depart slightly from my original variant in bf69b00e (not seen in @ap's version 9171033a8 as he modified that part), but the difference is obscure enough to effectively not matter.

TLDR: If the XS part ships as-is I won't be unhappy

P.S. Elaborating on the difference: it has to do with whether arguments to `S::N::subname` will or will not normalize `'` to `::`. My original test expected the literal `'` to remain in the newly created `GV`/`GV->STASH` `NAME`. @Leont himself alluded to this being consistent with perl itself:
```
#web-simple_20160619.log: <leont> It seems that setting the stash to «foo'bar» doesn't translate the ' to ::, possible I should compensate in SN
```
In any case - as I said earlier this is a very obscure corner case, and as long as `S::N` continues to do one thing consistently, it is immaterial which of the variants is being settled upon. Following @ap's test variant is fine by me.

**Comment by ap** [comment]
Well, the only way the semantics make sense to me is that the sub name must roundtrip unmolested. As far as Unicode characters are concerned, they must obviously go in and come back out the same.

I confess that I did not realise that Perl preserves the form of package separator. If this is in fact so, faithful roundtripping would dictate that the module ought preserve it as well. But if it really is the case, then Perl seems to be normalising the package separator in any circumstance I can think of where the distinction might make a difference (notably `caller`, natch). So assuming this is actually the case, it seems suboptimal but fine to me to ship without being able to preserve the separator and just mention it in LIMITATIONS.

Is it the case, though? All my attempts to find a `'` somewhere in some guts data structure have come up empty. Anywhere I even see a package separator stored, it seems to have been normalised to `::` ahead of time, as best I can tell.

**Comment by ribasushi** [comment]
> I confess that I did not realise that Perl preserves the form of package separator.

It does (afaict) during direct stash manipulation, but **does not** on `eval` (I suppose on the parser level). Hence why my 2c are "it's probably fine either way, as long as it is consistent for both `GV->NAME` and `GV->STASH->NAME`".

**Comment by Leont** [comment]
> Elaborating on the difference: it has to do with whether arguments to S::N::subname will or will not normalize ' to ::. My original test expected the literal ' to remain in the newly created GV/GV->STASH NAME. @Leont himself alluded to this being consistent with perl itself:

I think that normalizing to :: is the only sane thing possible to do towards users. Also because when you don't normalize it's the first use of the stash that matters, so if you would set the subname of two functions to `foo'bar'baz` and `foo::bar::quz`, the order would matter and one would not be what you expected.

**Comment by ribasushi** [comment]
> it's the first use of the stash that matters

This is something I was not aware of (even though it makes perfect sense in retrospect). In light of that: what you did is indeed **the only correct** way forward, and I withdraw [what I wrote in the P.S. above](https://github.com/p5sagit/Sub-Name/pull/8#issuecomment-227120151) entirely.

**Comment by karenetheridge** [comment]
This has been released as 0.16-TRIAL. Thanks all!
